doc/sphinx/addendum/ring.rst

```diff
@@ -1,736 +1,737 @@
-\end{coq_example}
-\begin{coq_eval}
-Abort.
-\end{coq_eval}
-\begin{coq_example}
-Goal forall a b:Z, 2*a*b = 30 ->
-        (a+b)^2 = a^2 + b^2 + 30.
-\end{coq_example}
-\begin{coq_example}
-intros a b H; ring [H].
-\end{coq_example} 
-\begin{coq_eval}
-Reset Initial.
-\end{coq_eval}
-
-\begin{Variants}
-  \item {\tt ring [\term$_1$ {\ldots} \term$_n$]} decides the equality of two
-    terms modulo ring operations and rewriting of the equalities
-    defined by \term$_1$ {\ldots} \term$_n$. Each of \term$_1$
-    {\ldots} \term$_n$ has to be a proof of some equality $m = p$,
-    where $m$ is a monomial (after ``abstraction''),
-    $p$ a polynomial and $=$ the corresponding equality of the ring structure.
-
-  \item {\tt ring\_simplify [\term$_1$ {\ldots} \term$_n$] $t_1 \ldots t_m$ in 
-{\ident}}
-     performs the simplification in the hypothesis named {\tt ident}.
-\end{Variants}
-
-\Warning \texttt{ring\_simplify \term$_1$; ring\_simplify \term$_2$} is
-not equivalent to \texttt{ring\_simplify \term$_1$ \term$_2$}. In the
-latter case the variables map is shared between the two terms, and
-common subterm $t$ of \term$_1$ and \term$_2$ will have the same
-associated variable number. So the first alternative should be
-avoided for terms belonging to the same ring theory.
-
-
-\begin{ErrMsgs}
-\item \errindex{not a valid ring equation}
-  The conclusion of the goal is not provable in the corresponding ring
+normalizing is automatically done by the tactic. So you can just
+forget this paragraph and use the tactic according to your intuition.
+
+Concrete usage in Coq
+--------------------------
+
+The ``ring`` tactic solves equations upon polynomial expressions of a ring
+(or semi-ring) structure. It proceeds by normalizing both hand sides
+of the equation (w.r.t. associativity, commutativity and
+distributivity, constant propagation, rewriting of monomials) and
+comparing syntactically the results.
+
+``ring_simplify`` applies the normalization procedure described above to
+the terms given. The tactic then replaces all occurrences of the terms
+given in the conclusion of the goal by their normal forms. If no term
+is given, then the conclusion should be an equation and both hand
+sides are normalized. The tactic can also be applied in a hypothesis.
+
+The tactic must be loaded by ``Require Import Ring``. The ring structures
+must be declared with the ``Add Ring`` command (see below). The ring of
+booleans is predefined; if one wants to use the tactic on |nat| one must
+first require the module ``ArithRing`` exported by ``Arith``); for |Z|, do
+``Require Import ZArithRing`` or simply ``Require Import ZArith``; for |N|, do
+``Require Import NArithRing`` or ``Require Import NArith``.
+
+
+.. example::
+
+  .. coqtop:: all
+
+    Require Import ZArith.
+    Open Scope Z_scope.
+    Goal forall a b c:Z, 
+        (a + b + c) ^ 2 = 
+         a * a + b ^ 2 + c * c + 2 * a * b + 2 * a * c + 2 * b * c.
+    intros; ring.
+    Goal forall a b:Z, 
```

Comment:
It would be better to add `Abort.` here as was originally the case.

Reply:
I think the infrastructure does `Abort` if you pass the right options to `..coqtop`. I don't think it is a good idea to keep it in the text, since it does not illustrate anything.

Reply:
Even if there is no new `..coqtop` in between?? Because let me remind you that nested lemmas are accepted in Coq and it looks very much like a nested lemma to me.

Reply:
Sorry, I hadn't noticed it was in the same domain.